akka-stream/src/main/scala/akka/stream/SslTlsOptions.scala

```diff
@@ -0,0 +1,255 @@
+ * each other. The difference between two half-duplex connections in opposite
+ * directions and a full-duplex connection is that the underlying transport
+ * is shared in the latter and tearing it down will end the data transfer in
+ * both directions.
+ *
+ * When integrating a full-duplex transport medium that does not support
+ * half-closing (which means ending one direction of data transfer without
+ * ending the other) into a stream topology, there can be unexpected effects.
+ * Feeding a finite Source into this medium will close the connection after
+ * all elements have been sent, which means that possible replies may not
+ * be received in full. To support this type of usage, the sending and
+ * receiving of data on the same side (e.g. on the [[Client]]) need to be
+ * coordinated such that it is known when all replies have been received.
+ * Only then should the transport be shut down.
+ *
+ * To support these scenarios it is recommended that the full-duplex
+ * transport integration is configurable in terms of termination handling,
+ * which means that the user can optionally suppress the normal (closing)
+ * reaction to completion or cancellation events, as is expressed by the
+ * possible values of this type:
+ *
+ *  - [[EagerClose]] means to not ignore signals
+ *  - [[IgnoreCancel]] means to not react to cancellation of the receiving
+ *    side unless the sending side has already completed
+ *  - [[IgnoreComplete]] means to not react to the completion of the sending
+ *    side unless the receiving side has already canceled
+ *  - [[IgnoreBoth]] means to ignore the first termination signal—be that
+ *    cancellation or completion—and only act upon the second one
+ */
+sealed abstract class TLSClosing {
+  def ignoreCancel: Boolean
+  def ignoreComplete: Boolean
+}
+object TLSClosing {
+  /**
+   * Java API: obtain the [[EagerClose]] singleton value.
+   */
+  def eagerClose: TLSClosing = EagerClose
+  /**
+   * Java API: obtain the [[IgnoreCancel]] singleton value.
+   */
+  def ignoreCancel: TLSClosing = IgnoreCancel
+  /**
+   * Java API: obtain the [[IgnoreComplete]] singleton value.
+   */
+  def ignoreComplete: TLSClosing = IgnoreComplete
+  /**
+   * Java API: obtain the [[IgnoreBoth]] singleton value.
+   */
+  def ignoreBoth: TLSClosing = IgnoreBoth
+}
+
+/**
+ * see [[TLSClosing]]
+ */
+sealed abstract class EagerClose extends TLSClosing {
+  override def ignoreCancel = false
+  override def ignoreComplete = false
+}
+case object EagerClose extends EagerClose
+
+/**
+ * see [[TLSClosing]]
+ */
+sealed abstract class IgnoreCancel extends TLSClosing {
+  override def ignoreCancel = true
+  override def ignoreComplete = false
+}
+case object IgnoreCancel extends IgnoreCancel
+
+/**
+ * see [[TLSClosing]]
+ */
+sealed abstract class IgnoreComplete extends TLSClosing {
+  override def ignoreCancel = false
+  override def ignoreComplete = true
+}
+case object IgnoreComplete extends IgnoreComplete
```

Comment:
First, it is not protocol, but options, second it would end up as object-in-object which is bad for Java.